File: trunk/openCF.agent.ruby/src/Agent.rb
require('socket')
require('json')
require('logger')
require('yaml')
require('fileutils')
require('net/http')
require('src/protocol')


$logger = Logger.new(STDOUT)
$logger.level = Logger::Severity::DEBUG


$config = YAML.load_file(File.absolute_path('config/agent.ruby.yaml'))
$os = $config['os']
$version = $config['version']
$server = $config['server']
$port = $config['port']
$heartbeat_sleep = $config['heartbeat']
$agent_id = $config['hostname']

$logger.info("os=#{$port}")
$logger.info("version=#{$version}")
$logger.info("server=#{$server}")
$logger.info("port=#{$port}")

class Agent
  
  private
    @hostname = "localhost"
    @port = 6789
    @socket = nil
    @registered = false
  
  public
  
  def initialize(host, port)
    @hostname = host
    @port = port
    @mutex = Mutex.new
    @jobs = Hash.new
  end
  
  
  def connect()
    @socket = TCPSocket.open(@hostname, @port)
  end
  
  
  def disconnect()
    @socket.close
  end
  
  
  def send(valueHash)
    json = JSON.generate valueHash
    
    $logger.debug "writing bytes: #{json.bytesize}"
    $logger.debug "writing data:  #{json}"
    
    @mutex.synchronize do    
      @socket.write(Protocol::htonl(json.bytesize))
      @socket.write(json)
    end
    
    @socket.flush()
  end
  
  
  def read()
    begin
      loop {
        response = Protocol::ntohl(@socket.read(4))
        #bytes = [response].pack('N')
        $logger.debug "#{response} zu lesen"
        json = JSON.parse(@socket.read(response))
        $logger.debug "gelesen #{json}"
        
        handlePacket(json)
      }
    rescue IOError => e
      $logger.error "io error"
    end
  end
  
  
  def handlePacket(json)
     type = json[Protocol::Keys::TYPE]
     $logger.debug "type: #{type}"
     
     case type
     when Protocol::Type::AGENT_HELLO_RESPONSE
       $logger.debug "agent_hello_response"
       handleHelloResponse(json)
     when Protocol::Type::AUTOMATION_CONTROL
       $logger.debug "automation_control"
       handleAutomationControl(json)
     else
       $logger.debug "unexpected packet recieved"
       handleUnexpectedType(json)
     end
  end
  
  
  def handleHelloResponse(json)
    successfull = json[Protocol::Keys::SUCCESSFULL]
    return_code = json[Protocol::Keys::RETURN_CODE]
    message     = json[Protocol::Keys::MESSAGE]
    
    $logger.debug "message:     #{message}"
    $logger.debug "return_code: #{return_code}"
    $logger.debug "successfull: #{successfull}"
    
    if successfull
      $logger.info "successfully registered to server"
      @registered = true
    else
      $logger.warn "registration failed"
      disconnect()
    end
    
  end
  
  
  def handleAutomationControl(json)
    
    action = json[Protocol::Keys::AUTOMATION_ACTION]
    id = json[Protocol::Keys::AUTOMATION_ID]
    $logger.debug "action: #{action}"
    
    case action
    when Protocol::Action::START
      $logger.info "automation start action"
      Thread.abort_on_exception = true
      thread = Thread.new {
        $logger.info("job thread started")
        $logger.debug(@jobs)
        @jobs[id] = Hash.new
        $logger.debug(@jobs[id])
        send(Protocol::automation_status(id, Protocol::Status::PREPARING, "prepareing"))
        ret = loadJob(json)
        if (ret == true)
          send(Protocol::automation_status(id, Protocol::Status::PREPARED, "prepared"))
        else
          send(Protocol::automation_status(id, Protocol::Status::PREPARE_FAILED, "prepare_failed"))
        end
        send(Protocol::automation_status(id, Protocol::Status::STARTED, "started"))
        Dir.chdir("executing/#{id}") do
          IO.popen("#{@jobs[id]['descriptor']['command']}", "r+")  do |pipe|
            pipe.sync = true
            while str = pipe.gets
                puts str
                send (Protocol::automation_status(id, Protocol::Status::TALKING, str))
              end
          end
        end
        removeJobDir(id)
        send(Protocol::automation_status(id, Protocol::Status::FINISHED, "finished"))
      }
    when Protocol::Action::STOP
      $logger.info "automation stop action"
    end
    
  end
  
  
  def createJobDir(id)
    begin
      Dir.chdir("executing") do
        puts Dir.pwd
        Dir::mkdir("#{id}")
      end
      return true
    rescue
      $logger.warn("job dir already exists")
      return false
    end
  end
  
  
  def removeJobDir(id)
    FileUtils.rm_rf("executing/#{id}")
  end
  
  
  def loadJob(json)
    url = json[Protocol::Keys::REPOSTITORY_URL]
    descriptor = json[Protocol::Keys::AUTOMATION_DESCRIPTOR]
    id = json[Protocol::Keys::AUTOMATION_ID]

    descriptorUri = URI(url + descriptor)
    
    $logger.debug("url: #{url}")
    $logger.debug("descriptor: #{descriptor}")
    $logger.debug("id: #{id}")
    
    $logger.debug("descriptorUri: #{descriptorUri}")
    
    resp = Net::HTTP.get(descriptorUri)
    
    $logger.debug("loaded data: #{resp}")
    
    descriptor = JSON.parse(resp)
    
    $logger.info("descriptor loaded: #{descriptor}")
    
    @jobs[id]['descriptor'] = descriptor
    
    uri = URI(url + descriptor['file'])
      
    $logger.debug("job uri: #{uri}")
    
    Net::HTTP.start(uri.host, uri.port) do |http|
      request = Net::HTTP::Get.new uri.request_uri
    
      createJobDir(id)
      
      http.request request do |response|
      case response
        when Net::HTTPOK
          Dir.chdir "executing/#{id}" do
            open "#{descriptor['file']}", 'w' do |io|
              response.read_body do |chunk|
                io.write chunk
              end
            end
          end
          return true
        when Net::HTTPNotFound
          $logger.warn("job not found")
          return false
        end
      end
    end
  end
  
  
  def handleUnexpactedType(json)
    $logger.warn "unexpected type: #{json}"
  end
  
  
  def to_s
    "Agent [hostname=#{@hostname}, port=#{@port}]"
  end
  
  
end

#{type:13, agent_id:["ruby_agent"], automation_action:"start","repository_url":"http://localhost:8080/jobs/", "automation_descriptor":"test.json"}
$hn = Socket.gethostname
if ($agent_id != nil)
  $hn = $agent_id
end
$logger.info("hostname=#{$hn}")


agent = Agent.new($server, $port)
agent.connect

thread = Thread.new{agent.read()}

agent.send(Protocol::agent_hello($hn, $version, $os))
t = Thread.new begin
  loop {
    $logger.debug("sleeping #{$heartbeat_sleep * 60} seconds")
    sleep($heartbeat_sleep * 60);
    agent.send(Protocol::heartbeat());
  }
end

thread.join()

agent.disconnect()

File: openCF.agent.perl/agent.pl
use strict;
use warnings;

use JSON;
use IO::Socket;
use Config::Properties;
use Log::Log4perl qw(get_logger :levels);


Log::Log4perl->init("log4perl.properties");

my $logger = Log::Log4perl::get_logger();

package Protocol::Types;
	sub INVALID					{return -1;}
	sub AGENT_HELLO 			{return  1;}
	sub AGENT_HELLO_RESPONSE  	{return  2;}
	sub AGENT_HEARTBEAT       	{return  0;}
	sub AUTOMATION_CONTROL    	{return 13;}
	sub AUTOMATION_STATUS     	{return 20;}

#package Protocol::Keys;
#	sub TYPE 				  {return "type";}
#	sub SUCCESSFULL           {return "successfull";}
#	sub RETURN_CODE           {return "return_code";}
#	sub MESSAGE               {return "message";}
#	sub AGENT_ID              {return "agent_id";}
#	sub AGENT_VERSION         {return "agent_version";}
#	sub AGENT_PLATTFORM       {return "agent_plattform";}
#	sub AGENT_ENCODING        {return "agent_encoding";}
#	sub AUTOMATION_STATUS     {return "automation_status";}
#	sub AUTOMATION_ACTION     {return "automation_action";}
#	sub AUTOMATION_ID         {return "automation_id";}
#	sub AUTOMATION_MESSAGE    {return "automation_message";}
#	sub AUTOMATION_DESCRIPTOR {return "automation_descriptor";}
#	sub AUTOMATION_PARAMETER  {return "automation_parameter";}
#	sub REPOSTITORY_URL       {return "repository_url";}
#	sub LOCAL_TIME            {return "local_time";}



package Connection;

sub new {
	$logger->trace("new");
	my ($type) = $_[0];
	my ($this) = {host => $_[1], port => $_[2]};

	$logger->debug("host: $_[1]");
	$logger->debug("port: $_[2]");

	my $socket;
	
	bless($this, $type);
	return($this)
}

sub connect {
	$logger->trace("connect");
	my ($this) = $_[0];
	
	$logger->info("connect to $this->{host}:$this->{port} via tcp");
	
	my $socket = IO::Socket::INET->new(
		PeerAddr => $this->{host},
		PeerPort => $this->{port},
		Proto => 'tcp') or die "Couldn't connect to Server\n";
	
	$this->{socket} = $socket;
}

sub readPacket {
	$logger->trace("readPacket");
	
	my ($this) = $_[0];
	my $recv_data;
	my $send_data;
	my $size = 4;
	
    $this->{socket}->read($recv_data, $size) or die "cant read size";
    $size = unpack ('N', $recv_data);
    $logger->debug("read size: ".$size."\n");
    $this->{socket}->read($recv_data, $size) or die "cant read data";
    $logger->debug("read data: ".$recv_data."\n");
    
    my $json = JSON->new->allow_nonref;
    
  	$this->handlePacket($json->decode($recv_data));
}

sub handlePacket {
	$logger->trace("handlePacket");
	
	my ($this) = $_[0];
	my ($packet) = $_[1];
	
	my $type = $packet->{type};
	
	$logger->debug("type: $type");
	
	if ($type eq Protocol::Types::AGENT_HELLO_RESPONSE) {
		$logger->info("agent hello response");
		handleAgentHelloResponse($packet);
	} elsif ($type eq Protocol::Types::AUTOMATION_CONTROL) {
		$logger->info("automation control");
		handleAutomationControl($packet);
	}
}

sub handleAgentHelloResponse {
	$logger->trace("handleAgentHelloResponse");
	my ($packet) = @_;
	
	my $message = $packet->{message};
	my $successfull = $packet->{successfull};
	my $return_code = $packet->{return_code};
	
	if ($successfull) {
		$logger->debug("login successfull");
	} elsif (not $successfull) {
		$logger->warning("login not successfull");
		$logger->info("message: $message");
		exit Protocol::Types::AGENT_HELLO_RESPONSE;
	}
}

sub handleAutomationControl {
	$logger->trace("handleAutomationControl");
	my ($packet) = @_;
	
}

sub sendPacket {
	$logger->trace("sendPacket");
	my ($this) = $_[0];
	my ($scalar) = $_[1];
	my $json = JSON->new->allow_nonref;
 
	my $pretty_printed = $json->encode($scalar);
	
	my $size = length($pretty_printed);

	$logger->debug("send size: ".$size."\n");
	$logger->debug("send data: ".$pretty_printed."\n");
	
	$this->{socket}->send(pack 'N', $size) or die "cant send size";
	$this->{socket}->send($pretty_printed) or die "cant send data";
}

  

package main;

open (my $fh, '<', 'agent.properties') or die ("unable to open configuration file");
my $properties = Config::Properties->new();
$properties->load($fh);
my $version = $properties->getProperty("version");
my $server = $properties->getProperty("server");
my $os = $properties->getProperty("os");
my $port = $properties->getProperty("port");
my $hostname = $properties->getProperty("hostname");

$logger->debug("version:  $version");
$logger->debug("server:   $server");
$logger->debug("os:       $os");
$logger->debug("port:     $port");
$logger->debug("hostname: $hostname");


$logger->debug("server: $server");
$logger->debug("port: $port");


my $connection = Connection->new($server, $port);
$connection->connect();
my $helloPacket = {
	type 			=> 1,
	agent_encoding 	=> "json",
	agent_id 		=> $hostname,
	agent_version 	=> $version,
	agent_plattform	=> $os
};
$connection->sendPacket($helloPacket);

while (1) {
	$logger->debug("start waiting for Packet");
	$connection->readPacket();
}
    
